src/neurostatslib/utils.py

```diff
@@ -4,44 +4,263 @@
-    """Parameters
+    Check that all arrays in iterable has the expected dimensionality.
+
+    Parameters
     ----------
-    basis_matrix : (n_basis_funcs, window_size)
-        Matrix holding 1d basis functions.
-    time_series : (n_neurons, n_timebins)
-        Matrix holding multivariate time series.
+    iterable :
+    Array-like object containing numpy or jax.numpy NDArrays.
+    expected_dim :
+    Number of expected dimension for the NDArrays.
 
     Returns
     -------
-    convolved_series : (n_neurons, n_basis_funcs, n_timebins - window_size + 1)
-        Result of convolution between all pairs of
-        features and basis functions.
+    True if all the arrays has the expected number of dimension, False otherwise.
+    """
+    return not any(array.ndim != expected_dim for array in iterable)
 
-    Notes
-    -----
-    For example, ``time_series`` could be a matrix of spike counts with
-    ``n_neurons`` neurons and ``n_timebins`` timebins, and ``basis_matrix``
-    could be a matrix of ``n_basis_funcs`` temporal basis functions with a
-    window size of ``window_size``.
 
+def convolve_1d_trials(
+    basis_matrix: ArrayLike,
+    time_series: Union[Iterable[NDArray], NDArray, Iterable[jnp.ndarray], jnp.ndarray],
+) -> List[jnp.ndarray]:
+    """
+    Convolve trial time series with a basis matrix.
+
+    This function checks if all trials have the same duration. If they do, it uses a fast method
+    to convolve all trials with the basis matrix at once. If they do not, it falls back to convolving
+    each trial individually.
+
+    Parameters
+    ----------
+    basis_matrix :
+        The basis matrix with which to convolve the trials. Shape (window_size, n_basis_funcs).
+    time_series :
+        The time series of trials to convolve with the basis matrix. It should be a list of 2D arrays,
+        where each array represents a trial and its second dimension matches the first dimension
+        of the basis_matrix. Each trial has shape (n_timebins_trial, n_neurons).
+
+    Returns
+    -------
+    :
+        The convolved trials. It is a list of 3D arrays, where each array represents a convolved trial.
+        Each element of the list will have shape (n_timebins_trial - window_size - 1, n_neurons, n_basis_funcs).
+
+    Raises
+    ------
+    ValueError
+        - If basis_matrix is not a 2D array-like object.
+        - If trials_time_series is not an iterable of 2D array-like objects.
+        - If trials_time_series contains empty trials.
+        - If the number of time points in each trial is less than the window size.
+    """
+
+    basis_matrix = jnp.asarray(basis_matrix)
+    # check input size
+    if basis_matrix.ndim != 2:
+        raise ValueError("basis_matrix must be a 2 dimensional array-like object.")
+    elif 0 in basis_matrix.shape:
+        raise ValueError(
+            "Empty basis_matrix provided. "
+            f"The shape of basis_matrix is {basis_matrix.shape}!"
+        )
```

Comment:
Does this come up often?

Reply:
It should not, but I see somebody filtering data with the wrong parameters and ending up with an empty basis. very rare probably